### PR TITLE
network: add notary request payload

### LINKED ISSUE #1546: P2P Notary request payload
Implement notary request payload (https://github.com/neo-project/neo/issues/1573#issuecomment-704874472).

### PR DESCRIPTION
**Pull request by AnnaShaleva**
Close #1546.

### CODE REVIEW

File: pkg/core/mempool/notary_request_pool.go [hunk 1]
```diff
@@ -0,0 +1,146 @@
...
+
+	n := sort.Search(len(mp.verifiedPayloads), func(n int) bool {
+		// TODO: compare fallback transactions?
+		return r.MainTransaction.NetworkFee-mp.verifiedPayloads[n].MainTransaction.NetworkFee > 0
+	})
+
+	// We've reached our capacity already.
+	if len(mp.verifiedPayloads) == mp.capacity {
+		// Less prioritized than the least prioritized we already have, won't fit.
+		if n == len(mp.verifiedPayloads) {
```
Comment (roman-khimov) [inline pkg/core/mempool/notary_request_pool.go:69]: I'm thinking about making it simpler, like FIFO. It opens up some opportunity for mischief (`defaultP2PNotaryRequestPayloadPoolSize × NotaryServiceFeePerKey` GAS and you've filled the queue, but this GAS will burn in `MaxNotValidBeforeDelta` blocks freeing up the queue again, but it simplifies things also (allowing to drop sorted slice completely)). Though current scheme is susceptible to this kind of attack also, I can add any kind of fee to the main transaction that won't ever be completed.

File: pkg/core/native/policy.go [hunk 2]
```diff
@@ -249,6 +253,15 @@ func (p *Policy) GetMaxVerificationGas(_ dao.DAO) int64 {
 	return defaultMaxVerificationGas
 }
 
+// GetMaxNotValidBeforeDelta returns the maximum allowed difference between NotValidBefore
+// and deposit expiration height for the fallback transaction.
+func (p *Policy) GetMaxNotValidBeforeDelta(_ dao.DAO) int64 {
```
Comment (roman-khimov) [inline pkg/core/native/policy.go:258]: There has to be some committee-signed setter for this.

File: pkg/network/payload/notary_request.go [hunk 3]
```diff
@@ -0,0 +1,147 @@
...
+	"github.com/nspcc-dev/neo-go/pkg/crypto/hash"
+	"github.com/nspcc-dev/neo-go/pkg/io"
+	"github.com/nspcc-dev/neo-go/pkg/util"
+)
+
+// P2PNotaryRequest contains main and fallback transactions for the Notary service.
+type P2PNotaryRequest struct {
+	MainTransaction     *transaction.Transaction
+	FallbackTransaction *transaction.Transaction
+	// TODO: Network?
```
Comment (roman-khimov) [inline pkg/network/payload/notary_request.go:16]: Implicit one, like with every other payload signed.

File: pkg/network/server.go [hunk 4]
```diff
@@ -678,6 +687,20 @@ func (s *Server) handleTxCmd(tx *transaction.Transaction) error {
...
+// handleP2PNotaryRequestCmd process received P2PNotaryRequest payload.
+func (s *Server) handleP2PNotaryRequestCmd(r *payload.P2PNotaryRequest) error {
+	if !s.chain.P2PSigExtensionsEnabled() {
+		return errors.New("P2PNotaryRequestCMD was received, but P2PSignatureExtensions are disabled")
+	}
+	if err := s.chain.PoolP2PNotaryRequestPayload(r); err != nil {
+		return err
+	}
+
+	msg := NewMessage(CMDInv, payload.NewInventory(payload.P2PNotaryRequestType, []util.Uint256{r.Hash()})) // TODO: retransmission?
```
Comment (roman-khimov) [inline pkg/network/server.go:699]: Same as for mempool.

**Review by roman-khimov — CHANGES_REQUESTED** [review]
One thing really worrying me here is tying request pool to the `Blockchain`. I'd really like to keep it in the `network.Server` somewhere.

Another big topic is separation of concerns between payload/pool/service checks. On one hand, I'd like to minimize the burden on non-service nodes (theoretically we could allow exchanging abstract blobs across the network just checking for the signature and deposit), on the other broadcasting invalid (to be rejected by the service) requests doesn't help the network either.

File: pkg/network/payload/inventory.go [hunk 5]
```diff
@@ -20,21 +20,24 @@ func (i InventoryType) String() string {
...
 
 // List of valid InventoryTypes.
 const (
-	TXType        InventoryType = 0x2b
-	BlockType     InventoryType = 0x2c
-	ConsensusType InventoryType = 0x2d
+	TXType               InventoryType = 0x2b
+	BlockType            InventoryType = 0x2c
+	ConsensusType        InventoryType = 0x2d
+	P2PNotaryRequestType InventoryType = 0x2e // TODO: another number, but not 0x2f
```
Comment (roman-khimov) [inline pkg/network/payload/inventory.go:40]: This BTW is reserved for state root messages in neo-project/neo#2079.

File: pkg/core/blockchain.go [hunk 6]
```diff
@@ -934,6 +947,11 @@ func (bc *Blockchain) GetGoverningTokenBalance(acc util.Uint160) (*big.Int, uint
 	return &neo.Balance, neo.LastUpdatedBlock
 }
 
+// GetDepositFor returns GAS amount deposited to Notary contract for the specified account.
+func (bc *Blockchain) GetDepositFor(acc util.Uint160) *big.Int {
+	return bc.contracts.Notary.BalanceOf(bc.dao, acc)
```
Comment (fyrchik) [inline pkg/core/blockchain.go:952]: I see how dependency on native contracts may prevent this from being done outside of `core`. May be we can do something like what was done in #1427.
E.g. updating `GetBalance()` function during `OnPersistEnd()`.

File: pkg/network/payloadpool/notary_request_pool.go [hunk 7]
```diff
@@ -0,0 +1,239 @@
...
+
+	newItem := item{
+		payload:    r,
+		blockStamp: bc.BlockHeight(),
+	}
+	mp.verifiedMap[r.Hash()] = r
+	if len(mp.verifiedHashes) == mp.capacity {
+		delete(mp.verifiedMap, mp.verifiedHashes[0].payload.Hash())
+		copy(mp.verifiedHashes, mp.verifiedHashes[1:])
+		mp.verifiedHashes[mp.capacity-1] = newItem
```
Comment (roman-khimov) [inline pkg/network/payloadpool/notary_request_pool.go:117]: But why? We can just `return ErrOOM` here.

File: pkg/network/payloadpool/notary_request_pool.go [hunk 8]
```diff
@@ -0,0 +1,239 @@
...
+type item struct {
+	payload    *payload.P2PNotaryRequest
+	blockStamp uint32
+}
+
+// NotaryRequestPool stores P2PNotaryRequest payloads.
+type NotaryRequestPool struct {
+	lock           sync.RWMutex
+	verifiedMap    map[util.Uint256]*payload.P2PNotaryRequest
+	verifiedHashes []item
```
Comment (roman-khimov) [inline pkg/network/payloadpool/notary_request_pool.go:34]: And you don't need it now.

File: pkg/network/notary_request_pool.go [hunk 9]
```diff
@@ -0,0 +1,244 @@
+package network
```
Comment (roman-khimov) [inline pkg/network/notary_request_pool.go:1]: I remember this pool in a separate package, why do we need it in the `network`? It can be `network/notarypool` or something like that.

File: pkg/network/payload/notary_request.go [hunk 10]
```diff
@@ -0,0 +1,144 @@
...
+
+func (r *P2PNotaryRequest) isValid() error {
+	nKeysMain := r.MainTransaction.GetAttributes(transaction.NotaryAssistedT)
+	if len(nKeysMain) == 0 {
+		return errors.New("main transaction should have NotaryAssisted attribute")
+	}
+	if nKeysMain[0].Value.(*transaction.NotaryAssisted).NKeys == 0 {
+		return errors.New("main transaction should have NKeys > 0")
+	}
+	// TODO: check "either doesn't have all witnesses attached (in this case none of them can be multisignature), or it only has a partial multisignature, only one of the two is allowed "
```
Comment (roman-khimov) [inline pkg/network/payload/notary_request.go:115]: Maybe we can leave it to the module. The payload as it is now is guaranteed to have valid fallback transaction and random node on the network should probably only care about that.

File: pkg/core/blockchainer/policer.go [hunk 11]
```diff
@@ -0,0 +1,17 @@
...
+import (
+	"math/big"
+
+	"github.com/nspcc-dev/neo-go/pkg/util"
+)
+
+// Policer is an interface that abstracts the implementation of policy methods.
+type Policer interface {
+	GetDepositBalance(acc util.Uint160) *big.Int
+	GetDepositExpiration(acc util.Uint160) uint32
```
Comment (roman-khimov) [inline pkg/core/blockchainer/policer.go:12]: This isn't part of the policer, it's something different.

File: pkg/network/server.go [hunk 12]
```diff
@@ -682,6 +697,25 @@ func (s *Server) handleTxCmd(tx *transaction.Transaction) error {
...
+func (s *Server) handleP2PNotaryRequestCmd(r *payload.P2PNotaryRequest) error {
+	if !s.chain.P2PSigExtensionsEnabled() {
+		return errors.New("P2PNotaryRequestCMD was received, but P2PSignatureExtensions are disabled")
+	}
+	if err := s.chain.RunWithCurrentState(func() error {
+		return s.notaryRequestPool.VerifyAndPoolPayload(s.chain, r)
+	}); err != nil {
+		return err
+	}
+	go s.broadcastP2PNotaryRequestPayload(r)
```
Comment (roman-khimov) [inline pkg/network/server.go:710]: Doing it for each payload is not a good idea, it can be synchronous (or batched as transactions are).

**Review by roman-khimov — CHANGES_REQUESTED** [review]
Looks good, `VerifyAndPoolPayload` for state-dependent things and `isValid` for state-independent. All the logic is in place and just needs to be moved a little.

File: pkg/network/server.go (hunk 12 above)
Comment (roman-khimov) [reply pkg/network/server.go:710]: Maybe just synchronous for now, we can add batching later if needed.

File: pkg/network/notary_request_pool.go (hunk 9 above)
Comment (AnnaShaleva) [reply pkg/network/notary_request_pool.go]: This was mostly done for testing. Now `NotaryPool` is in `network/notarypool`, but we need a Blockchainer for primitive tests. At the same time we cannot move `notarypool_test.go` to the `network` package because we need access to private fields of the pool. As a result, I had to copy `network.testchain` to the `notarypool` package.

File: pkg/core/blockchainer/policer.go (hunk 11 above)
Comment (AnnaShaleva) [reply pkg/core/blockchainer/policer.go]: Moved out of the Policer to the Blockchainer.

File: pkg/network/server.go (hunk 12 above)
Comment (AnnaShaleva) [reply pkg/network/server.go]: Made it synchronous.

File: pkg/network/notarypool/notary_pool.go [hunk 13]
```diff
@@ -0,0 +1,246 @@
...
+	}
+	if err := bc.VerifyWitness(payer, r.FallbackTransaction, &r.FallbackTransaction.Scripts[1], mp.policer.GetMaxVerificationGAS()); err != nil {
+		return fmt.Errorf("bad fallback transaction witness #1: %s", err.Error())
+	}
+	nvbFallback := r.FallbackTransaction.GetAttributes(transaction.NotValidBeforeT)[0].Value.(*transaction.NotValidBefore).Height
+	maxNVBDelta := mp.policer.GetMaxNotValidBeforeDelta()
+	if bc.BlockHeight()+maxNVBDelta < nvbFallback {
+		return fmt.Errorf("fallback transaction should have NotValidBefore not more then %d, got %d", bc.BlockHeight()+maxNVBDelta, nvbFallback)
+	}
+	if nvbFallback+maxNVBDelta < r.FallbackTransaction.ValidUntilBlock { // TODO: <= ?
```
Comment (AnnaShaleva) [inline pkg/network/notarypool/notary_pool.go:99]: There's one last question, don't we want to make here `<=`?

File: pkg/core/blockchain.go (hunk 6 above)
Comment (AnnaShaleva) [reply pkg/core/blockchain.go:959]: We have `(*NotaryPool).PostBlock` now, which can make use of `Blockchainer`. Do we still need this updating function?

File: pkg/network/notarypool/notary_pool.go (hunk 13 above)
Comment (roman-khimov) [reply pkg/network/notarypool/notary_pool.go:99]: For chain height H the transaction is valid if `NVB == H` and not valid if `VUB == H`. Here we want transaction to be valid for `maxNVBDelta` number of blocks, I think we're getting this property with `<`, so it should be fine.

File: pkg/network/notary_request_pool.go (hunk 9 above)
Comment (roman-khimov) [reply pkg/network/notary_request_pool.go]: Move it out to `internal` then, we have a number of testing packages there shared among several real packages and some `fakechain` is appropriate there.

File: pkg/core/blockchain_test.go [hunk 14]
```diff
@@ -1437,3 +1439,173 @@ func TestRemoveUntraceable(t *testing.T) {
 	require.NoError(t, err)
 	require.Len(t, b.Transactions, 0)
 }
+
+func TestVerifyAndPoolPayload(t *testing.T) {
+	// we need a proper Blockchainer for this test, so we have to keep it in core package
```
Comment (roman-khimov) [inline pkg/core/blockchain_test.go:1444]: One more thing for #1472. This really is a `TestChain`, btw.

File: pkg/core/blockchain.go [hunk 15]
```diff
@@ -934,6 +954,16 @@ func (bc *Blockchain) GetGoverningTokenBalance(acc util.Uint160) (*big.Int, uint
 	return &neo.Balance, neo.LastUpdatedBlock
 }
 
+// GetDepositBalance returns Notary deposit amount for the specified account.
+func (bc *Blockchain) GetDepositBalance(acc util.Uint160) *big.Int {
```
Comment (roman-khimov) [inline pkg/core/blockchain.go:958]: `GetNotaryBalance`? `Deposit` is too generic for `Blockchain`.

File: pkg/core/mempool/mem_pool.go [hunk 16]
```diff
@@ -403,6 +441,21 @@ func (mp *Pool) TryGetValue(hash util.Uint256) (*transaction.Transaction, bool)
 	return nil, false
 }
 
+// TryGetData returns data associated with the specified transaction if it exists in the memory pool.
+func (mp *Pool) TryGetData(hash util.Uint256) (interface{}, bool) {
+	mp.lock.RLock()
+	defer mp.lock.RUnlock()
+	if _, ok := mp.verifiedMap[hash]; ok {
+		for _, itm := range mp.verifiedTxes {
```
Comment (roman-khimov) [inline pkg/core/mempool/mem_pool.go:449]: Very inefficient.

File: pkg/core/mempool/mem_pool.go [hunk 17]
```diff
@@ -303,7 +341,7 @@ func (mp *Pool) removeInternal(hash util.Uint256, feer Feer) {
 // RemoveStale filters verified transactions through the given function keeping
 // only the transactions for which it returns a true result. It's used to quickly
 // drop part of the mempool that is now invalid after the block acceptance.
-func (mp *Pool) RemoveStale(isOK func(*transaction.Transaction) bool, feer Feer) {
+func (mp *Pool) RemoveStale(isOK func(*transaction.Transaction, interface{}) bool, feer Feer) {
```
Comment (roman-khimov) [inline pkg/core/mempool/mem_pool.go:344]: I don't think you need this `interface{}`, to make a decision you already have everything needed in the transaction.

File: pkg/core/mempool/mem_pool.go (hunk 16 above)
Comment (AnnaShaleva) [reply pkg/core/mempool/mem_pool.go]: Replaced by `sort.Search`.

File: pkg/core/mempool/mem_pool.go (hunk 17 above)
Comment (AnnaShaleva) [reply pkg/core/mempool/mem_pool.go:321]: We need `data`  here because it contains MainTx. And `isOK` for payload includes MainTx checks.

Comment (roman-khimov) [reply pkg/core/mempool/mem_pool.go:321]: You're only using `ValidUntilBlock` (but it's the same as in fallback transaction) and `Hash()` (but it's in the `Conflicts` attribute of fallback).

Comment (AnnaShaleva) [reply pkg/core/mempool/mem_pool.go:321]: But fallback transaction might have several `Conflicts` attributes (currently we check only the presence of mainTx.Hash in the list of fallbackTx's conflicts attributes). We can either restrict the number of conflicts attributes for the fallback transaction by 1 or reserve position (e.g. first place in the attributes array) for the attribute with the mainTx hash.

Comment (roman-khimov) [reply pkg/core/mempool/mem_pool.go:321]: But if any of them is to enter the chain the fallback can no longer be valid, so you need to check for all of them.

It at the same time shows another potential problem, if I'm to create a fallback conflicting with main tx as well as some $OTHER_TX, I can send this $OTHER_TX right before NVB making the request invalid. Though I'll have to pay for this $OTHER_TX, it can be less than what fallback tx pays with its fees. So we can require fallback to have exactly one conflicting attribute.

Comment (roman-khimov) [reply pkg/core/mempool/mem_pool.go:321]: At least for now, to simplify things. Maybe it's not that big of an issue, but it's safer to add more restrictions at the moment while keeping our main use cases covered.

File: pkg/core/mempool/mem_pool.go [hunk 18]
```diff
@@ -256,7 +269,7 @@ func (mp *Pool) Add(t *transaction.Transaction, fee Feer) error {
 		}
 	}
 	// we already checked balance in checkTxConflicts, so don't need to check again
-	mp.tryAddSendersFee(pItem.txn, fee, false)
+	mp.tryAddSendersFee(t, fee, false)
```
Comment (roman-khimov) [inline pkg/core/mempool/mem_pool.go:272]: Some unnecessary change.

File: pkg/core/blockchain.go [hunk 19]
```diff
@@ -1290,6 +1320,9 @@ func (bc *Blockchain) verifyAndPoolTx(t *transaction.Transaction, pool *mempool.
 		return fmt.Errorf("%w: %v", ErrPolicy, err)
 	}
 	size := t.Size()
+	if isPartialTx { // then it's a fallback transaction from NotaryRequest with missing Notary witness
+		size += 66 // InvocationScript: 2 bytes for PUSHDATA1 + 64 bytes for signature; VerificationScript: 0 bytes (empty script)
```
Comment (roman-khimov) [inline pkg/core/blockchain.go:1324]: What if we're to require invocation script to be "almost" correct? In that it will `PUSHDATA1(0x000...)`, which will make tx size correct and allow to run verification script as usual (even though it will fail).

File: pkg/core/mempool/mem_pool.go [hunk 20]
```diff
@@ -403,6 +420,27 @@ func (mp *Pool) TryGetValue(hash util.Uint256) (*transaction.Transaction, bool)
...
+func (mp *Pool) TryGetData(hash util.Uint256) (interface{}, bool) {
+	mp.lock.RLock()
+	defer mp.lock.RUnlock()
+	if tx, ok := mp.verifiedMap[hash]; ok {
+		itm := item{txn: tx}
+		n := sort.Search(len(mp.verifiedTxes), func(n int) bool {
+			return itm.CompareTo(mp.verifiedTxes[n]) >= 0
+		})
+		if n < len(mp.verifiedTxes) {
+			for i := n; i < len(mp.verifiedTxes); i++ { // items may have equal priority, so `n` is the left bound of the items which are as prioritized as the desired `itm`.
```
Comment (roman-khimov) [inline pkg/core/mempool/mem_pool.go:433]: You can break out of the loop if `itm.CompareTo() != 0`.

File: pkg/core/mempool/mem_pool.go [hunk 21]
```diff
@@ -403,6 +420,27 @@ func (mp *Pool) TryGetValue(hash util.Uint256) (*transaction.Transaction, bool)
...
 }
 
+// TryGetData returns data associated with the specified transaction if it exists in the memory pool.
+func (mp *Pool) TryGetData(hash util.Uint256) (interface{}, bool) {
+	mp.lock.RLock()
+	defer mp.lock.RUnlock()
+	if tx, ok := mp.verifiedMap[hash]; ok {
+		itm := item{txn: tx}
+		n := sort.Search(len(mp.verifiedTxes), func(n int) bool {
+			return itm.CompareTo(mp.verifiedTxes[n]) >= 0
```
Comment (roman-khimov) [inline pkg/core/mempool/mem_pool.go:430]: `== 0`?

Comment (AnnaShaleva) [reply pkg/core/mempool/mem_pool.go:430]: But this is a sort.Search, and `==0` is not a monotonous function. Am I missing something?

Comment (roman-khimov) [reply pkg/core/mempool/mem_pool.go:430]: Yeah, you're right, I'm looking at C# code too much these days probably.

File: pkg/core/mempool/mem_pool.go (hunk 20 above)
Comment (AnnaShaleva) [reply pkg/core/mempool/mem_pool.go:433]: I thought about it, but there might be a case when a lot of items are iterated before this condition is satisfied. The other way to stop the cycle is to search the right bound of items which are as prioritized as the desired one by using `sort.Search` with `itm.CompareTo(mp.verifiedTxes[n]) > 0` condition. Which is more preferable?

Comment (roman-khimov) [reply pkg/core/mempool/mem_pool.go:433]: > a case when a lot of items are iterated before this condition is satisfied

But it's `<= len(mp.verifiedTxes)` anyway. There are from 0 to `len(mp.verifiedTxes) - n` transactions satisfying `itm.CompareTo() == 0`.

File: pkg/core/blockchain.go (hunk 19 above)
Comment (AnnaShaleva) [reply pkg/core/blockchain.go]: Done, and also added several checks to NotaryRequest.DecodeBinary. We have to check the length of both invocation and verification scripts for dummy witness, and also have to check the PUSHDATA1 prefix in order to match the desired verification price exactly.

File: pkg/core/blockchain.go [hunk 22]
```diff
@@ -1757,3 +1811,41 @@ func (bc *Blockchain) newInteropContext(trigger trigger.Type, d dao.DAO, block *
...
 }
+
+// RegisterPostBlock appends provided function to the list of functions which should be run after new block
+// is stored.
+func (bc *Blockchain) RegisterPostBlock(f func(blockchainer.Blockchainer, *mempool.Pool, *block.Block)) {
+	bc.postBlock = append(bc.postBlock, f)
+}
+
+// RunWithCurrentState allows to call f under the Blockchain lock.
+func (bc *Blockchain) RunWithCurrentState(f func() error) error {
```
Comment (roman-khimov) [inline pkg/core/blockchain.go:1822]: I don't see it being used now, seems like `PoolTxWithData` completely replaced it.

File: pkg/core/mempool/mem_pool.go [hunk 23]
```diff
@@ -94,6 +96,12 @@ func (p item) CompareTo(otherP item) int {
 	return int(p.txn.NetworkFee - otherP.txn.NetworkFee)
 }
 
+// FeePerByte returns item's FeePerByte with respect to additionalSizeBytes value.
+func (p item) FeePerByte(additionalSizeBytes int) int64 {
```
Comment (roman-khimov) [inline pkg/core/mempool/mem_pool.go:100]: No longer needed.

File: pkg/network/server.go [hunk 24]
```diff
@@ -687,11 +708,62 @@ func (s *Server) handleTxCmd(tx *transaction.Transaction) error {
...
+		return errors.New("P2PNotaryRequestCMD was received, but P2PSignatureExtensions are disabled")
+	}
+	if s.verifyAndPoolPayload(r) == RelaySucceed {
+		s.broadcastP2PNotaryRequestPayload(nil, r)
+	}
+	return nil
+}
+
+// verifyAndPoolNotaryRequest verifies NotaryRequest payload and adds it to the payload mempool.
+func (s *Server) verifyAndPoolPayload(r *payload.P2PNotaryRequest) RelayReason {
```
Comment (roman-khimov) [inline pkg/network/server.go:728]: `verifyAndPoolNotaryRequest`? `Payload` is just about everything here.

**Review by roman-khimov — COMMENTED** [review]
Looks good.

**Review by roman-khimov — APPROVED** [review]
